nano/core_test/network.cpp

```diff
@@ -1423,3 +1423,165 @@ TEST (bootstrap, keepalive)
+		ASSERT_EQ (nano::process_result::progress, node->ledger.process (transaction, receive2).code);
+
+		// Check confirmation heights of all the accounts are uninitialized (0),
+		// as we have any just added them to the ledger and not processed any live transactions yet.
+		nano::account_info account_info;
+		ASSERT_FALSE (node->store.account_get (transaction, nano::test_genesis_key.pub, account_info));
+		ASSERT_EQ (0, account_info.confirmation_height);
+		ASSERT_FALSE (node->store.account_get (transaction, key1.pub, account_info));
+		ASSERT_EQ (0, account_info.confirmation_height);
+		ASSERT_FALSE (node->store.account_get (transaction, key2.pub, account_info));
+		ASSERT_EQ (0, account_info.confirmation_height);
+		ASSERT_FALSE (node->store.account_get (transaction, key3.pub, account_info));
+		ASSERT_EQ (0, account_info.confirmation_height);
+	}
+
+	// The nodes process a live receive which propagates across to all accounts
+	auto receive3 = std::make_shared<nano::receive_block> (open3.hash (), send6.hash (), key3.prv, key3.pub, system.work.generate (open3.hash ()));
+
+	for (auto & node : system.nodes)
+	{
+		node->process_active (receive3);
+		node->block_processor.flush ();
+
+		system.deadline_set (10s);
+		while (true)
+		{
+			auto transaction = node->store.tx_begin_read ();
+			if (node->ledger.block_confirmed (transaction, receive3->hash ()))
+			{
+				break;
+			}
+
+			ASSERT_NO_ERROR (system.poll ());
+		}
+
+		nano::account_info account_info;
+		auto & store = node->store;
+		auto transaction = node->store.tx_begin_read ();
+		ASSERT_FALSE (store.account_get (transaction, nano::test_genesis_key.pub, account_info));
+		ASSERT_EQ (4, account_info.confirmation_height);
+		ASSERT_EQ (4, account_info.block_count);
+		ASSERT_FALSE (store.account_get (transaction, key1.pub, account_info));
+		ASSERT_EQ (2, account_info.confirmation_height);
+		ASSERT_EQ (3, account_info.block_count);
+		ASSERT_FALSE (store.account_get (transaction, key2.pub, account_info));
+		ASSERT_EQ (3, account_info.confirmation_height);
+		ASSERT_EQ (4, account_info.block_count);
+		ASSERT_FALSE (store.account_get (transaction, key3.pub, account_info));
+		ASSERT_EQ (2, account_info.confirmation_height);
+		ASSERT_EQ (2, account_info.block_count);
+
+		// The accounts for key1 and key2 have 1 more block in the chain than is confirmed.
+		// So this can be rolled back, but the one before that cannot. Check that this is the case
+		{
+			auto transaction = node->store.tx_begin_write ();
+			ASSERT_FALSE (node->ledger.rollback (transaction, node->latest (key2.pub)));
+			ASSERT_FALSE (node->ledger.rollback (transaction, node->latest (key1.pub)));
+		}
+		{
+			// These rollbacks should fail
+			auto transaction = node->store.tx_begin_write ();
+			ASSERT_TRUE (node->ledger.rollback (transaction, node->latest (key2.pub)));
+			ASSERT_TRUE (node->ledger.rollback (transaction, node->latest (key1.pub)));
```

Comment:
Could we add check for latest hashes probably? send4 & send6 I guess

Reply:
Good idea, I'm testing rollbacks on the latest blocks of the other accounts and a few other cemented blocks.